### DISCUSSION TITLE
plugins.twitch: hls-live-edge documentation clarification

### DISCUSSION THREAD

**Discussion by Bertaz** [streamlink/streamlink#5138] [Q&A] [answered]
### Description

Twitch plugin documentation about [Low latency streaming](https://streamlink.github.io/cli/plugins/twitch.html#low-latency-streaming) referring to `--hls-live-edge` says "Setting the value to `1` is not advised due to how prefetching works."
cli documentation about [--hls-live-edge](https://streamlink.github.io/cli.html#cmdoption-hls-live-edge) say nothing about a value of `1` and instead explain that lower values decrease latency increasing risk of buffering. This was and still is my understanding of live-edge.

Does a value of `1` not work correctly with twitch hls segments prefetching? Does it give somehow a higher latency than `2`?

**Comment by bastimeyer** [comment, marked as answer]
HLS streams are segmented streams where each segment gets concatenated in the final output. Each egment does have a certain length (or size), depending on the streaming provider and its implementation. Segments on Twitch usually are 2 seconds long, but this varies between different streams, whether they are actually low-latency streaming, and it also seems to depend on their individual contract with Twitch (there are some big/early/OG streamers with lower segment times).

The live-edge value simply sets the number of segments from the current live position in the stream where to begin the stream output upon the first playlist retrieval. Depending on the segment lengths and advertised playlist refresh time, this value needs to be adjusted to prevent buffering and to be able to be as close as possible to the actual live position.

The official HLS specification does not have support for actual low-latency streams, so over the years, custom solutions like LHLS and provider-specific stuff were implemented by certain streaming providers. A new HLS specification with proper low-latency support is on its way though, similar to these custom solutions, but it's not finalized yet, and using it is not advised, hence why it hasn't been implemented in Streamlink yet and why nobody is using it in production yet. That's why Twitch is (still) using a custom implementation.

This custom implementation adds two future-segment URLs without additional metadata to the playlist which can be requested by the HLS client ahead of time. Requests of those future segments will block until actual content arrives. The intention of this is that this reduces latency in-between the necessary playlist refreshes, and the segment download starts right away when content becomes available.

However, due to Streamlink's HLS implementation which doesn't just queue these requests ahead of time and treats them differently, it rather shifts the live-edge towards these future-segments, as they are treated as regular segments in the stream sequence (with metadata extrapolation from the regular segments). Since Twitch includes exactly two future-segments, reducing the live-edge value to 1 doesn't have any effect on the latency. What it can do though is increase buffering, either at the start of the stream or when the stream is not a low-latency stream, which can't be queried on any of their APIs and can only be figured out when parsing the playlist, when it's already too late. If the stream is not a LL stream and you've set the live-edge value to 1, then there's not enough time to refresh the playlist and queue the next segment.

So just leave the live-edge value untouched. If you want to make any further low-latency tweaks, then take a look at your video player's configuration and its own buffer.

There's a possibility that the low-latency implementation will change once the new revised HLS specification gets published, and this will most likely affect Twitch and Streamlink's custom TwitchHLS implementation as well. If that's the case, then the live-edge stuff will most likely change as well, but that's nothing an end-user needs to care about, especially now.

**Reply by Bertaz** [reply]
Thank you very much for the in depth explanation
